File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isLower, isSymbol, isDigit, isPunctuation, isSpace)
import Data.List (dropWhileEnd)

trim = dropWhileEnd isSpace . dropWhile isSpace

responseFor :: String -> String
responseFor xs
 | isCaps xs && (isQuestion xs) && (hasChars xs) = "Calm down, I know what I'm doing!"
 | (isCaps xs) && (hasChars xs) = "Whoa, chill out!"
 | isQuestion (trim xs) = "Sure."
 | all isSpace xs = "Fine. Be that way!"
 | otherwise = "Whatever."
 where
   hasChars = any (\x -> isUpper x || (isLower x))
   isQuestion xs = not (null xs) && (last xs) == '?'
   isCaps = all isUpperChar
   isUpperChar x = (isSymbol x) || (isDigit x) || (isUpper x) || (isPunctuation x) || (isSpace x)
   

File: solutions/scala/bob/1/src/main/scala/Bob.scala
object Bob:
  def isQuestion(str: String): Boolean = str.endsWith("?")
  def isYell(str: String): Boolean = str.toUpperCase == str && str.exists(_.isLetter)

  def response(prompt: String): String =
    prompt.trim match
      case str if str.isBlank => "Fine. Be that way!"
      case str if isQuestion(str) && isYell(str) => "Calm down, I know what I'm doing!"
      case str if isQuestion(str) => "Sure."
      case str if isYell(str)=> "Whoa, chill out!"
      case _ => "Whatever."
